File: python_port/ingestion/batch_process.py
"""Python port of matlab/ingestion/batchprocess.m.

Daily extractor for Nodo-equivalent data (1-day bins):
  - Scans *.bin recursively under a chosen source directory to find the
    calendar days actually present
  - For each day D: tStart = D 00:00:00, tEnd = D+1 00:00:00 (next day)
  - Calls load_nodo_data(tStart, tEnd, fsamp, rootDir)
  - Saves under <out-dir>/<DatiXX>/Nodo_<DatiXX>_yyyyMMdd_yyyyMMdd.pkl
  - Skips output files that already exist

Deviations from the MATLAB source (all deliberate, see README.md):
  - No uigetdir GUI picker; the source folder is a required CLI argument.
  - No hardcoded date-range filter (batchprocess.m had one baked in at the
    time this was ported); pass --start-date/--end-date to restrict it.
  - Output is pickled Python objects (list[dict]), not a MATLAB .mat file.
  - parfor is replaced by an optional ProcessPoolExecutor (--workers).
"""
from __future__ import annotations

import argparse
import pickle
import re
import sys
import traceback
from concurrent.futures import ProcessPoolExecutor, as_completed
from pathlib import Path
from typing import List, Optional

import numpy as np

if __package__:
    from .load_nodo_data import load_nodo_data
else:  # allow `python batch_process.py ...` as well as `-m python_port.ingestion.batch_process`
    sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent))
    from python_port.ingestion.load_nodo_data import load_nodo_data

_FNAME_RE = re.compile(
    r"^(?P<YYYY>\d{4})_(?P<MMDD>\d{4})(?P<HH>\d{2})(?P<MN>\d{2})(?P<SS>\d{2}).*?_(?P<kind>pjm|p)\.bin$"
)


def _file_end_time(name: str) -> Optional[np.datetime64]:
    m = _FNAME_RE.match(name)
    if not m:
        return None
    g = m.groupdict()
    return np.datetime64(
        f"{g['YYYY']}-{g['MMDD'][:2]}-{g['MMDD'][2:]}T{g['HH']}:{g['MN']}:{g['SS']}"
    )


def _dataset_tag(root_dir: Path) -> str:
    m = re.search(r"Dati\d+", str(root_dir))
    return m.group(0) if m else "DatiXX"


def _process_one_day(day_start: np.datetime64, root_dir: str, fsamp: float,
                      out_dir: str, dataset_tag: str) -> str:
    day_start = np.datetime64(day_start, "us")
    day_end = day_start + np.timedelta64(1, "D")
    out_path = Path(out_dir) / dataset_tag
    out_path.mkdir(parents=True, exist_ok=True)

    base_name = (f"Nodo_{dataset_tag}_"
                 f"{str(day_start)[:10].replace('-', '')}_"
                 f"{str(day_end)[:10].replace('-', '')}.pkl")
    out_file = out_path / base_name

    if out_file.is_file():
        return f"SKIP (exists): {out_file}"

    try:
        nodo = load_nodo_data(day_start, day_end, fsamp, root_dir)
        if not nodo:
            return f"EMPTY: {day_start} -> {day_end}"
        with open(out_file, "wb") as f:
            pickle.dump(nodo, f, protocol=pickle.HIGHEST_PROTOCOL)
        return f"OK: {out_file}"
    except Exception as exc:
        err_file = out_path / f"ERROR_{str(day_start)[:10].replace('-', '')}_{str(day_end)[:10].replace('-', '')}.txt"
        with open(err_file, "w", encoding="utf-8") as f:
            f.write(f"Error on day {day_start}-{day_end}\n")
            f.write(f"Source: {root_dir}\n")
            f.write(f"Message: {exc}\n\n")
            f.write(traceback.format_exc())
        return f"ERROR (see {err_file}): {exc}"


def run_batch(root_dir: Path, out_dir: Path, fsamp: float = 1,
              start_date: Optional[np.datetime64] = None,
              end_date: Optional[np.datetime64] = None,
              workers: int = 1) -> None:
    root_dir = Path(root_dir)
    dataset_tag = _dataset_tag(root_dir)

    all_bins = sorted(root_dir.rglob("*.bin"))
    if not all_bins:
        print(f"No *.bin files found under: {root_dir}")
        return

    end_times: List[np.datetime64] = []
    for p in all_bins:
        et = _file_end_time(p.name)
        if et is not None:
            end_times.append(et)
    if not end_times:
        print(f"No filenames matched the expected pattern under: {root_dir}")
        return

    end_times = np.array(end_times, dtype="datetime64[us]")
    if start_date is not None:
        end_times = end_times[end_times >= np.datetime64(start_date, "us")]
    if end_date is not None:
        end_times = end_times[end_times < np.datetime64(end_date, "us")]
    if end_times.size == 0:
        print(f"No files in the time range {start_date} to {end_date}.")
        return

    day_starts = np.unique(end_times.astype("datetime64[D]")).astype("datetime64[us]")
    n_days = len(day_starts)
    print(f"Found {n_days} day(s) with data in: {root_dir}")

    out_dir.mkdir(parents=True, exist_ok=True)

    if workers <= 1:
        for d in day_starts:
            print(_process_one_day(d, str(root_dir), fsamp, str(out_dir), dataset_tag))
    else:
        with ProcessPoolExecutor(max_workers=workers) as pool:
            futures = {
                pool.submit(_process_one_day, d, str(root_dir), fsamp, str(out_dir), dataset_tag): d
                for d in day_starts
            }
            completed = 0
            for fut in as_completed(futures):
                completed += 1
                print(f"[{completed}/{n_days}] {fut.result()}")

    print(f"Done. Processed {n_days} day(s).")


def main(argv=None):
    parser = argparse.ArgumentParser(description="Port of matlab/ingestion/batchprocess.m")
    parser.add_argument("root_dir", type=Path, help="Raw BIN folder, e.g. data/raw/Dati10")
    parser.add_argument("--out-dir", type=Path, default=None,
                         help="Output root (default: <repo_root>/data/interim/python_port -- "
                              "deliberately NOT data/interim/<DatiXX> directly, since that is "
                              "where matlab/ingestion/batchprocess.m writes its own .mat output "
                              "using the same folder-naming convention; keeping this port's .pkl "
                              "output in its own subfolder avoids the two ever colliding or being "
                              "mistaken for one another)")
    parser.add_argument("--fsamp", type=float, default=1,
                         help="40 selects the 40 Hz 'HP' packet layout; anything else selects "
                              "the fixed 1.62181 Hz 'LP' layout (matches loadNodoData.m's FCAMP switch).")
    parser.add_argument("--start-date", type=str, default=None, help="Inclusive, e.g. 2026-01-01")
    parser.add_argument("--end-date", type=str, default=None, help="Exclusive, e.g. 2026-02-18")
    parser.add_argument("--workers", type=int, default=1, help="1 = serial; >1 uses a process pool")
    args = parser.parse_args(argv)

    if args.out_dir is None:
        if __package__:
            from ..paths import get_paths
        else:
            from python_port.paths import get_paths
        args.out_dir = get_paths().interim / "python_port"

    start_date = np.datetime64(args.start_date) if args.start_date else None
    end_date = np.datetime64(args.end_date) if args.end_date else None

    run_batch(args.root_dir, args.out_dir, fsamp=args.fsamp,
              start_date=start_date, end_date=end_date, workers=args.workers)


if __name__ == "__main__":
    main()

File: matlab/ingestion/batchprocess.m
%% Daily extractor for Nodo data (1-day bins, parallelized)
% - Prompts for a source directory (for example, data/raw/DatiXX)
% - Scans *.bin recursively to find actual days present
% - For each day D: tStart = D 00:00:00, tEnd = D+1 00:00:00 (next day)
% - Calls loadNodoData(tStart,tEnd,rootDir)
% - Saves under the configured data/interim directory as:
%     Nodo_DatiXX_yyyymmdd_yyyymmdd.mat
% Notes:
%   * Uses parfor to process days independently
%   * Skips output files that already exist
%   * Uses -v7.3 for large structs

clc; clear; close all;
thisFile = mfilename('fullpath');
projectRoot = fileparts(fileparts(fileparts(thisFile)));
addpath(fullfile(projectRoot, 'config'));
paths = matlab_paths();
fsamp = 1;

%% 1) Select source directory (rootDir) that contains the raw *.bin files
rootDir = uigetdir(paths.raw, 'Select a raw BIN folder such as data/raw/DatiXX');
if isequal(rootDir,0)
    disp('User canceled.'); return;
end
rootDir = string(rootDir);

% Try to extract "DatiXX" from the selected path for naming
tok = regexp(rootDir, 'Dati\d+', 'match', 'once');
if isempty(tok); tok = "DatiXX"; end  
datasetTag = string(tok);

%% 2) Discover all *.bin and their timestamps (yyyy_MMddHHmmss_*_(pjm|p).bin)
S = dir(fullfile(rootDir, '**', '*.bin'));
if isempty(S)
    warning('No *.bin files found under: %s', rootDir);
    return;
end
names   = string({S.name}).';
folders = string({S.folder}).';
paths   = fullfile(folders, names);

% Same robust filename parser as loadNodoData
expr = "^(?<YYYY>\d{4})_(?<MMDD>\d{4})(?<HH>\d{2})(?<MN>\d{2})(?<SS>\d{2}).*?_(?<kind>pjm|p)\.bin$";
tokn = regexp(names, expr, 'names');

isMatch = ~cellfun('isempty', tokn);
if ~any(isMatch)
    warning('No filenames matched the expected pattern under: %s', rootDir);
    return;
end
tokn   = tokn(isMatch);
paths  = paths(isMatch);
names  = names(isMatch);

YYYY = string(cellfun(@(t) t.YYYY, tokn, 'UniformOutput', false));
MMDD = string(cellfun(@(t) t.MMDD, tokn, 'UniformOutput', false));
HH   = string(cellfun(@(t) t.HH,   tokn, 'UniformOutput', false));
MN   = string(cellfun(@(t) t.MN,   tokn, 'UniformOutput', false));
SS   = string(cellfun(@(t) t.SS,   tokn, 'UniformOutput', false));
% kind = string(cellfun(@(t) t.kind, tokn, 'UniformOutput', false));  % not needed here

dtstr = YYYY + extractBetween(MMDD,1,2) + extractBetween(MMDD,3,4) + HH + MN + SS;
fileEndTime = datetime(dtstr, 'InputFormat','yyyyMMddHHmmss');  % tz-naive

%% >>> NEW BLOCK: limit to recent files only (from 1 Oct for 1 month) <<<
tMin = datetime(2026,01,01);          % start date inclusive
tMax = datetime(2026,02,18);           % one-month window: [1 Dec, 31 Dec)

inRange = (fileEndTime >= tMin) & (fileEndTime < tMax);

if ~any(inRange)
    warning('No files in the time range %s to %s.', ...
            datestr(tMin), datestr(tMax));
    return;
end

% Keep only files in the desired date range
fileEndTime = fileEndTime(inRange);
tokn        = tokn(inRange);
paths       = paths(inRange);
names       = names(inRange);
%% <<< END NEW BLOCK >>>

% Build daily bins actually present in data
dayStarts = unique(dateshift(fileEndTime, 'start', 'day'));  % midnight anchors
nDays = numel(dayStarts);
fprintf('Found %d day(s) with data in: %s\n', nDays, rootDir);


if nDays==0
    warning('No valid day boundaries could be formed.'); return;
end

%% 3) Prepare output root and per-dataset subfolder
OUT_ROOT = paths.interim;
if ~isfolder(OUT_ROOT), mkdir(OUT_ROOT); end
OUT_DIR = fullfile(OUT_ROOT, datasetTag);
if ~isfolder(OUT_DIR), mkdir(OUT_DIR); end

%% 4) Spin up a parallel pool (force process-based, v7.3-safe)
try
    % 1) Kill any existing pool (especially threads-based)
    p = gcp('nocreate');
    if ~isempty(p)
        try
            if isprop(p,'Cluster') && isprop(p.Cluster,'Type')
                fprintf('Existing pool type: %s — deleting...\n', p.Cluster.Type);
            end
        end
        delete(p);
        p = [];
    end

    % 2) Force a process pool (works across releases)
    % R2023b+ supports this:
    started = false;
    try
        parpool("Processes");   % explicitly request processes
        started = true;
    catch
        % Older MATLAB: ensure 'local' profile is active then start process pool
        try
            parallel.defaultClusterProfile('local');
        catch
            % ignore if not available
        end
        parpool('local');       % process-backed 'local' cluster
        started = true;
    end

    % 3) Verify and print
    if started
        p = gcp;  % should exist now
        poolType = '';
        try
            poolType = p.Cluster.Type;
        end
        fprintf('Parallel pool started. Cluster Type = %s, Workers = %d\n', poolType, p.NumWorkers);

        % Guard: if we still somehow got threads, stop and run serial to avoid save errors
        if strcmpi(poolType, 'threads')
            warning('Process pool was requested but a threads pool is active. Closing to avoid v7.3 save errors.');
            delete(p);
        end
    end

catch ME
    warning(ME.identifier, '%s. Continuing serially...', ME.message);
end

%% 5) Progress tracking (works in both serial and parallel contexts)
completed = 0;
total     = nDays;
t0        = tic;
if exist('parallel.pool.DataQueue','class')==8
    dq = parallel.pool.DataQueue;
    afterEach(dq, @(~) fprintf('Progress: %d/%d days (%.1f%%)\n', ...
        min(completed+1,total), total, 100*min(completed+1,total)/total));
else
    dq = [];
end

%% 6) Process each day → call loadNodoData(tStart,tEnd,rootDir) and save
% One-day span: [tStart, tEnd), naming uses yyyyMMdd_yyyyMMdd (start and next day)
parfor d = 1:nDays
    tStart = dayStarts(d);
    tEnd   = tStart + days(1);      % next midnight (inclusive in your function's filter)

    % Output filename: Nodo_DatiXX_yyyymmdd_yyyymmdd.mat
    baseName = "Nodo_" + datasetTag + "_" + string(tStart,'yyyyMMdd') + "_" + string(tEnd,'yyyyMMdd') + ".mat";
    outFile  = fullfile(OUT_DIR, baseName);

    % Skip if already exists
    if isfile(outFile)
        if ~isempty(dq), send(dq, 1); end
        continue;
    end

    try
        % Core call
        
        Nodo = loadNodoData(tStart, tEnd, fsamp, rootDir);

        % If empty, you may skip writing to avoid clutter
        if isempty(Nodo)
            % optionally, write a small marker MAT stating "empty day"
            % save(outFile, 'Nodo', '-v7.3');
            if ~isempty(dq), send(dq, 1); end
            continue;
        end

        % Save (v7.3 handles large structs safely)
        % --- inside parfor, replace the save(...) block with:
        try
            if ~isfolder(fileparts(outFile)), mkdir(fileparts(outFile)); end
            M = matfile(outFile, 'Writable', true);   % creates v7.3 .mat if new
            M.Nodo = Nodo;                             % write variable
        catch ME
            errFile = fullfile(OUT_DIR, "ERROR_" + string(tStart,'yyyyMMdd') + "_" + string(tEnd,'yyyyMMdd') + ".txt");
            fid = fopen(errFile, 'w');
            if fid > 0
                fprintf(fid, 'Error on day %s–%s\nSource: %s\nMessage: %s\n', ...
                    string(tStart,'yyyy-MM-dd'), string(tEnd,'yyyy-MM-dd'), ME.identifier, ME.message);
                if ~isempty(ME.stack)
                    for sidx = 1:numel(ME.stack)
                        fprintf(fid, '  at %s (line %d)\n', ME.stack(sidx).name, ME.stack(sidx).line);
                    end
                end
                fclose(fid);
            end
        end

    catch ME
        % Log an error file next to outputs for post-mortem
        errFile = fullfile(OUT_DIR, "ERROR_" + string(tStart,'yyyyMMdd') + "_" + string(tEnd,'yyyyMMdd') + ".txt");
        fid = fopen(errFile, 'w');
        if fid > 0
            fprintf(fid, 'Error on day %s–%s\nSource: %s\nMessage: %s\n', ...
                string(tStart,'yyyy-MM-dd'), string(tEnd,'yyyy-MM-dd'), ME.identifier, ME.message);
            if ~isempty(ME.stack)
                for sidx = 1:numel(ME.stack)
                    fprintf(fid, '  at %s (line %d)\n', ME.stack(sidx).name, ME.stack(sidx).line);
                end
            end
            fclose(fid);
        end
    end

    if ~isempty(dq), send(dq, 1); end
end

elapsedSec = toc(t0);
fprintf('Done. Processed %d day(s). Elapsed: %.1f s\n', nDays, elapsedSec);
